### PR TITLE
allowing reporting of models using redis

### PR DESCRIPTION
**Pull request by miltontony**

### CODE REVIEW

File: ummeli/reporting/helpers.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+def get_object_vote_key(obj, report_key_field):
+    return '-'.join([obj.content_type.app_label,
+        obj.content_type.model,
+        obj.slug,
+        report_key_field
+    ])
+
+def get_user_vote_key(user, obj, report_key_field):
+    return '-'.join([user.username,
```
Comment (smn) [inline ummeli/reporting/helpers.py:14]: What's the likelihood of there being a `-` in the username?

File: ummeli/reporting/helpers.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        get_object_vote_key(obj, report_key_field)
+    ])
+
+def get_object_votes(obj, report_key_field):
+    redis_server = redis.StrictRedis(host='localhost', port=6379, db=0)
```
Comment (smn) [inline ummeli/reporting/helpers.py]: perhaps move these **kwargs to a settings file?

File: ummeli/reporting/helpers.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    key = get_object_vote_key(obj, report_key_field)
+    votes = redis_server.get(key) or 0
+    return int(votes)
+
+def has_voted(user, obj, report_key_field):
+    redis_server = redis.StrictRedis(host='localhost', port=6379, db=0)
```
Comment (smn) [inline ummeli/reporting/helpers.py]: Is there a way to not have to connect every time?

File: ummeli/reporting/helpers.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    key = get_user_vote_key(user, obj, report_key_field)
+    return redis_server.get(key)
```
Comment (smn) [inline ummeli/reporting/helpers.py]: I was expecting an `exists(key)` call here. Why are you doing a `get(key)`?

File: ummeli/reporting/helpers.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+def vote(user, obj, report_key_field):
+    redis_server = redis.StrictRedis(host='localhost', port=6379, db=0)
+    print redis_server
```
Comment (smn) [inline ummeli/reporting/helpers.py]: stray!

File: ummeli/reporting/helpers.py (hunk 1 above)
Comment (miltontony) [inline ummeli/reporting/helpers.py:14]: Would it matter? It would still create a valid key

File: ummeli/reporting/helpers.py (hunk 2 above)
Comment (miltontony) [inline ummeli/reporting/helpers.py]: How do you guys handle this for vumi?

**Comment by smn** [comment]
How does this tie into the Django admin?

File: ummeli/reporting/helpers.py (hunk 4 above)
Comment (miltontony) [inline ummeli/reporting/helpers.py]: ah yes. Fixed.

**Comment by miltontony** [comment]
At the moment it doesn't tie into Djando Admin, but from our earlier discussion on a service oriented approach, we wouldn't want this to tie into django admin. If this was hosted on a separate machine, we'd have to write a reporting interface.

File: ummeli/reporting/helpers.py (hunk 2 above)
Comment (smn) [inline ummeli/reporting/helpers.py]:
``` python

from django.conf import settings

redis_server = redis.StrictRedis(**settings.REDIS_SETTINGS)
```

File: ummeli/reporting/helpers.py (hunk 1 above)
Comment (smn) [inline ummeli/reporting/helpers.py:14]: True, but splitting it reliably at a later stage can become burdensome. You decide if you think you need that or not.

**Comment by miltontony** [comment]
Alternatively, we can write a Django Admin view that looks at all the models that are tagged with `can_report` and connects to the service and retrieve the values for each model

File: ummeli/reporting/helpers.py (hunk 1 above)
Comment (miltontony) [inline ummeli/reporting/helpers.py:14]: What would be a better implementation? Using a different separator?

Comment (smn) [inline ummeli/reporting/helpers.py:14]: Possibly false sense of security but perhaps something less likely to show up in a username? `::`, `:`, `/` ?

**Comment by smn** [comment]
:+1: in general, my other comments are insignificant in the light of our ideas of growing this into a dedicated service.

**Comment by miltontony** [comment]
would you rather I created a `jmbo-reporting` repo instead?

**Comment by miltontony** [comment]
hmm...`jmbo-reporting` sounds like something that generates report metrics etc

**Comment by smn** [comment]
Nah, let's not create a new repo just yet. :+1: and :cake: for landing this here now.

**Pull request merged**